src/jupyterhub_config/custom_kube_spawner.py

```diff
@@ -0,0 +1,345 @@
+        shutil.copy2(bashrc_tmpl, user_dir / '.bashrc.tmpl')
+        shutil.copy2(bash_profile_tmpl, user_dir / '.bash_profile.tmpl')
+
+        bashrc_dest = user_dir / '.bashrc'
+        bash_profile_dest = user_dir / '.bash_profile'
+
+        if not bashrc_dest.exists():
+            self.log.info(f'Creating .bashrc file for {user_dir}')
+            shutil.copy2(bashrc_tmpl, bashrc_dest)
+
+        if not bash_profile_dest.exists():
+            self.log.info(f'Creating .bash_profile file for {user_dir}')
+            shutil.copy2(bash_profile_tmpl, bash_profile_dest)
+
+    def _ensure_user_jupyter_directory(self, user_dir: Path):
+        """
+        Create the user's Jupyter directory and subdirectories if they do not exist. And set the
+        environment variables for Jupyter to use these directories.
+        """
+
+        if not user_dir.exists():
+            raise ValueError(f'User directory {user_dir} does not exist')
+
+        jupyter_dir = user_dir / '.jupyter'
+        jupyter_runtime_dir = jupyter_dir / 'runtime'
+        juputer_data_dir = jupyter_dir / 'data'
+
+        jupyter_dir.mkdir(parents=True, exist_ok=True)
+        jupyter_runtime_dir.mkdir(parents=True, exist_ok=True)
+        juputer_data_dir.mkdir(parents=True, exist_ok=True)
+
+        # copy the jupyter_jupyter_ai_config.json file to the user's .jupyter directory
+        # ref: https://jupyter-ai.readthedocs.io/en/latest/users/index.html#configuring-as-a-config-file
+        jupyter_notebook_config = Path(os.environ['JUPYTERHUB_CONFIG_DIR']) / os.environ['JUPYTER_AI_CONFIG_FILE']
+        shutil.copy2(jupyter_notebook_config, jupyter_dir / os.environ['JUPYTER_AI_CONFIG_FILE'])
+
+        self.environment['JUPYTER_CONFIG_DIR'] = str(jupyter_dir)
+        self.environment['JUPYTER_RUNTIME_DIR'] = str(jupyter_runtime_dir)
+        self.environment['JUPYTER_DATA_DIR'] = str(juputer_data_dir)
+
+    def _ensure_virtual_environment(self, user_env_dir: Path):
+        """
+        Ensure the user's virtual environment exists. If it does not exist, it is
+        created with the system site-packages included.
+        """
+        if not user_env_dir.exists():
+            user_env_dir.mkdir(parents=True, exist_ok=True)
+            self.log.info(f'Creating virtual environment for {self.user.name}')
+            try:
+                # Create a virtual environment with system site-packages access
+                venv.create(env_dir=user_env_dir, system_site_packages=True, with_pip=True)
+            except Exception as e:
+                raise ValueError(f'Failed to create virtual environment for {self.user.name}: {e}') from e
+        else:
+            self.log.info(f'Reusing virtual environment for {self.user.name}')
+
+    def _configure_environment(self, user_dir: Path, user_env_dir: Path, username: str):
+        self.environment.update({key: str(value) for key, value in os.environ.items() if key not in self.environment})
```

Comment:
Is it possible to have non str values in os.environ?

Reply:
No. Kubenetes doesn't allow non-str env variables.

Reply:
What's the purpose of stringifying the value in that case?

Reply:
Just so that we still keep everything in case we need it.

Reply:
But if os.environ can't have non-str values why bother with the str()?

Reply:
So that we keep a copy of things in case we need to use it. I don't really get what you mean actually.

Reply:
The code is calling str() on the values of the os.environ dict. If the values of the os.environ dict are required to be strings already, it's a noop and there's no need for the str() call

Reply:
You can still set os.environ whatever you want. `self.environment` doesn't allow non-string values for KubeSpawner.

Reply:
:+1: